slasher/rpc/detect_update_min_max_span.go

```diff
@@ -0,0 +1,93 @@
+package rpc
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	ethpb "github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+// DetectAndUpdateMaxSpan is used to detect and update the max span of an incoming attestation.
+// logic is following the detection method designed by https://github.com/protolambda
+// from here: https://github.com/protolambda/eth2-surround/blob/master/README.md#min-max-surround
+func (ss *Server) DetectAndUpdateMaxSpan(ctx context.Context, source uint64, target uint64, validatorIdx uint64) (surroundedTargetEpoch uint64, err error) {
+	span := target - source + 1
+	if span > params.BeaconConfig().WeakSubjectivityPeriod {
+		return 0, fmt.Errorf("attestation detection supports only weak subjectivity period: %v target - source: %v > weakSubjectivityPeriod", params.BeaconConfig().WeakSubjectivityPeriod, span)
+	}
+	spanMap, err := ss.SlasherDB.ValidatorSpansMap(validatorIdx)
+	if err != nil {
+		return 0, errors.Wrapf(err, "could not retrieve span map for validatorIdx: %v", validatorIdx)
+	}
+	if spanMap.EpochSpanMap == nil {
```

Comment:
spanMap itself could be nil, no? That should also be checked. Maybe we can also do:

```go
if _, ok := spanMap.EpochSpanMap[source]; ok {
}
```